src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/LocalDB.Windows.cs

```diff
@@ -0,0 +1,222 @@
+            int sizeOfbuffer = localDBConnectionString.Capacity;
+            localDBStartInsanceFunc(localDbInstance, 0, localDBConnectionString, ref sizeOfbuffer);
+            return localDBConnectionString.ToString();
+        }
+
+        internal enum LocalDBErrorState
+        {
+            NO_INSTALLATION, INVALID_CONFIG, NO_SQLUSERINSTANCEDLL_PATH, INVALID_SQLUSERINSTANCEDLL_PATH, NONE
+        }
+
+        internal static uint MapLocalDBErrorStateToCode(LocalDBErrorState errorState)
+        {
+            switch (errorState)
+            {
+                case LocalDBErrorState.NO_INSTALLATION:
+                    return SNICommon.LocalDBNoInstallation;
+                case LocalDBErrorState.INVALID_CONFIG:
+                    return SNICommon.LocalDBInvalidConfig;
+                case LocalDBErrorState.NO_SQLUSERINSTANCEDLL_PATH:
+                    return SNICommon.LocalDBNoSqlUserInstanceDllPath;
+                case LocalDBErrorState.INVALID_SQLUSERINSTANCEDLL_PATH:
+                    return SNICommon.LocalDBInvalidSqlUserInstanceDllPath;
+                case LocalDBErrorState.NONE:
+                    return 0;
+                default:
+                    return SNICommon.LocalDBInvalidConfig;
+            }
+        }
+
+        /// <summary>
+        /// Loads the User Instance dll.
+        /// </summary>
+        private bool LoadUserInstanceDll()
+        {
+            // Check in a non thread-safe way if the handle is already set for performance.
+            if (_sqlUserInstanceLibraryHandle != null)
+            {
+                return true;
+            }
+
+            lock (this)
+            {
+                if(_sqlUserInstanceLibraryHandle !=null)
+                {
+                    return true;
+                }
+                //Get UserInstance Dll path
+                LocalDBErrorState registryQueryErrorState;
+
+                // Get the LocalDB instance dll path from the registry
+                string dllPath = GetUserInstanceDllPath(out registryQueryErrorState);
+
+                // If there was no DLL path found, then there is an error.
+                if (dllPath == null)
+                {
+                    SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.INVALID_PROV, 0, MapLocalDBErrorStateToCode(registryQueryErrorState), string.Empty);
+                    return false;
+                }
+
+                // In case the registry had an empty path for dll
+                if (string.IsNullOrWhiteSpace(dllPath))
+                {
+                    SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.INVALID_PROV, 0, SNICommon.LocalDBInvalidSqlUserInstanceDllPath, string.Empty);
+                    return false;
+                }
+
+                // Load the dll
+                SafeLibraryHandle libraryHandle = Interop.Kernel32.LoadLibraryExW(dllPath.Trim(), IntPtr.Zero, 0);
```

Comment:
The DLL path is being fetched from the registry. Registry modification is only allowed to the admins. 

The DLL path, is an absolute path so DLLs will load from an absolute path and the search order of dll loading will not come into play here.

@stephentoub  Does this answer the question? Let me know if I have missed out any aspects. 
Are you hinting at the application memory being modified to modify the DLL path during the load?